src/gpio.rs

```diff
@@ -1,4 +1,65 @@
 //! GPIO and pin configuration
+//!
+//! Interface for the input/output pins.
+//!
+//! `Gpio0` through `GpioNN` represent the pin peripherals of the ESP chip. You
+//! may think of pin peripherals as *physical* pins.
+//!
+//! Pin drivers are implemented through `PinDriver`. You may think of pin drivers as
+//! *logical* I/O pins. They implement the [`embedded_hal::digital::InputPin`](https://docs.rs/embedded-hal/latest/embedded_hal/digital/trait.InputPin.html)/[`OutputPin`](https://docs.rs/embedded-hal/latest/embedded_hal/digital/trait.OutputPin.html)
+//! traits, as well as their [`embedded_hal_async`](https://docs.rs/embedded-hal-async/latest/embedded_hal_async/) counterparts.
+//!
+//! The ESP architecture has a [I/O multiplexer](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#iomuxgpio),
+//! which means that (almost) any physical pin can be used for any logical
+//! function (i.e. GPIO, I2C, SPI, ADC, etc).
+//!
+//! Reusing a pin for several functions is possible but should be avoided (since
```

Comment:
I think this paragraph is not a correct statement. The type system makes sure that if you for example are giving out `&mut gpio0) than it will check it for you. So in that sense we do this complete shenanigans (compared to the underlying C api where you just give for example an arbitrary int number) so the user don't need to be careful here at all and nothing of that sort is discouraged . The compiler has you here! 

And sorry for the delay i think otherwise everything is fine and we should merge it then.